[Loysius Lam] iP

Blizzeracz opened the pull request:
### LoyBot

> “Anything that can go wrong will go wrong." -Murphy's Law

LoyBot frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST **_SUPER_** FAST to use

All you need to do is,
1. start the bot by typing `java -jar Duke.jar` in your CLI!
2. add your tasks.
3. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x]  Managing tasks

- [x] Managing deadlines 

- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Some basic LoyBot commands are
```
delete
todo
event from: "YYYY/MM/DD" to: "YYYY/MM/DD"
deadline by: "YYYY/MM/DD"
mark "task number"
unmark "task number"

```

Please support my [GitHub](https://github.com/Blizzeracz) !!! 😄 


![cute-shark-eating-bone-waving-hand-cartoon-vector-icon-illustration-animal-nature-isolated-flat_138676-6427](https://github.com/nus-cs2103-AY2324S1/ip/assets/111228229/81e86774-3104-4de8-82ac-af0091c66fb0)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Blizzeracz committed:
Change the name of the bot and giving an outro

Blizzeracz committed:
Change how the bot reacts to certain words, echoing or exiting
based off the keywords

Blizzeracz committed:
Add the ability to store whatever text entered by user and display them back to user when requested

Blizzeracz committed:
Add the ability to mark tasks as done and undone and create a new class called task

Blizzeracz committed:
Create subclasses under Task such as ToDos, Events and Deadlines. Clean up some of the code to make it more readable

Blizzeracz committed:
Create the input and expected text file for automated testing of the bot

Blizzeracz committed:
Create a new class DukeException for error handling. Use try catch block for exception handling as well

Blizzeracz committed:
Add support for deleting tasks from the list and change normal aray to arraylist

Blizzeracz committed:
Add a storage class which save the tasks and load it from the disk

Blizzeracz committed:
Merge branch 'branch-Level-7'

Blizzeracz committed:
Create a way for the bot to understand date format

Blizzeracz committed:
Merge branch 'branch-Level-8'

Blizzeracz committed:
Make the code more OOP

Blizzeracz committed:
Improve the OOP codes and debug the errors

Blizzeracz committed:
Divide the class into their respective packages for organisation

Blizzeracz committed:
Merge branch 'branch-A-Gradle'

Blizzeracz committed:
Set up gradle to clean, build and run Duke

Blizzeracz committed:
Add JUnit tests for Parser and Storage methods

Blizzeracz committed:
Edit gradle to automate deletion of files

Blizzeracz committed:
Add JavaDoc comments

Blizzeracz committed:
Add a find function for users to find a task

Blizzeracz committed:
Fix the merge conflict

Blizzeracz committed:
Fix all checkstyle issues in different class

Blizzeracz committed:
Change the checkStyle of the test files

jacksonyuan256 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,105 @@
+ */
+public class Storage {
+
+    private static final String FOLDER_PATH = "./data";
```
Good use of CAPITALISE naming.

jacksonyuan256 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,105 @@
+            }
+
+            Scanner scanner = new Scanner(file);
+            while (scanner.hasNextLine()) {
```
I am not entirely sure what is happening with the split and replace?

jacksonyuan256 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,217 @@
+     * @param ui    The Ui instance for user interaction.
+     * @throws DukeException If an error occurs during parsing or execution.
+     */
+    public static void parse(String input, TaskList tasks, Ui ui) throws DukeException {
```
Good use of switch statement.

jacksonyuan256 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,217 @@
+     * @return An EventTask created from the input.
+     * @throws DukeException If the command is not properly formatted.
+     */
+    public static EventTask parseEvent(String taskDetails, boolean isDone) throws DukeException {
```
Most of the method names are appropriately named, good job!

jacksonyuan256 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,76 @@
+     * Displays a welcome message when the program starts.
+     */
+    public void showWelcomeMessage() {
+        System.out.println("Hi! This is your AI assistant LoyBoy!");
```
You can store these strings as a final variable as you did in Storage. This is still okay though.

jacksonyuan256 reviewed (commented):
Overall, the code is easy to read, methods are well-named.

Blizzeracz committed:
Refactor and add JavaFX GUI

Blizzeracz committed:
Merge branch 'branch-GUI'

MagnificentCreature commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,72 @@
+
+    private DialogBox(String text, Image img) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(MainWindow.class.getResource("/view/DialogBox.fxml"));
```
Might you consider storing this as a constant? Since it is a static filepath

MagnificentCreature commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,52 @@
+    public void start(Stage stage) {
+        try {
+            // Load the FXML file for the main window
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
```
Similar to above, could this perhaps be a constant?

MagnificentCreature commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,55 @@
+
+    private Duke duke;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/DaUser.png"));
```
Similar to above

MagnificentCreature commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,213 @@
+            case "mark":
+                int markTask = parseMarkCommand(parts) - 1;
+                if (markTask < 0 || markTask >= tasks.getTotalTasks()) {
+                    return ui.showNoTaskFound();
```
I like the use of switch, but this if statement broke the "pattern" of calling and then returning a parsed command. Could this if statement perhaps be abstracted inside the parseMarkCommand and thrown as an error? Similar to how the delete command operates

MagnificentCreature commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,213 @@
+            case "list":
+                return tasks.listTask();
+            case "delete":
+                int task = parseDeleteCommand(parts);
```
I like the use of error handling abstracted inside this function!

MagnificentCreature commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,213 @@
+                return tasks.getTask(markTask).markTask();
+            case "unmark":
+                int unmarkTask = parseUnmarkCommand(parts) - 1;
+                if (unmarkTask < 0 || unmarkTask >= tasks.getTotalTasks()) {
```
similar to above

MagnificentCreature commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,213 @@
+     * @throws DukeException If an error occurs during parsing.
+     */
+    public static Task parseLoad(String taskType, String taskDetails, boolean isDone) throws DukeException {
+        if (taskType.equalsIgnoreCase("[T")) {
```
could this be a switch case?

MagnificentCreature commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,105 @@
+
+    private static final String FOLDER_PATH = "./data";
+
+    private static final String DATA_FILE_PATH = "./data/duke.txt";
```
As jackson has said, I like the use of constants for static filepaths

MagnificentCreature reviewed (commented):
Overall well written code with a few nits.

Blizzeracz committed:
Add Assertions for Error Handling

Some methods do not check if there is any programming errors and logic
mistakes.

By adding assertions, it can be use to 'assert' certain conditions hold
true with regards ti the code, identifiying bugs earky.

Methods that takes receives variables are being asserted such to ensure
that earlier methods catches any other runtime errors such as user
input.

Blizzeracz committed:
Enhance code quality

Some methods are too long and have different levels of abstraction while
some classes have dead codes and code duplication.

This makes the classes seem like they have a long line of code and
methods, making it uneasy to read.

Creating new methods and level of abstractions helps to reduce
duplicated codes and makes it neater. The dead codes are also deleted
making the source code more logical.

Blizzeracz committed:
Merge pull request #1 from Blizzeracz/branch-A-assertions

Add Assertions for Error Handling

Blizzeracz committed:
Merge pull request #2 from Blizzeracz/branch-A-CodeQuality

Enhance code quality

Blizzeracz committed:
Add a sort function that sorts the list based on date

Blizzeracz committed:
Set up continuous integration using GitHub Actions

Blizzeracz committed:
Tweak the padding, fonts and colors for the GUI to be more pleasing

Blizzeracz committed:
Elongate the gui

Blizzeracz committed:
Fix sort numbers

Blizzeracz committed:
Add the user guide

Blizzeracz committed:
Clean up files and ready for release

Blizzeracz committed:
Edit the style for most classes to pass checkstyle

Blizzeracz committed:
Fix build gradle file to work on other OS

Blizzeracz committed:
Improve the user guide
